sql/stats/join_test.go

```diff
@@ -0,0 +1,449 @@
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+)
+
+func TestBinMerge(t *testing.T) {
+	tests := []struct {
+		inp []sql.HistogramBucket
+		exp []sql.HistogramBucket
+	}{
+		{
+			inp: sql.Histogram{
+				&Bucket{RowCnt: 5, DistinctCnt: 5, BoundVal: sql.Row{2}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{2}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{3}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{4}, BoundCnt: 5},
+			},
+			exp: sql.Histogram{
+				&Bucket{RowCnt: 10, DistinctCnt: 5, BoundVal: sql.Row{2}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{3}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{4}, BoundCnt: 5},
+			},
+		},
+		{
+			inp: sql.Histogram{
+				&Bucket{RowCnt: 5, DistinctCnt: 10, BoundVal: sql.Row{2}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{3}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{3}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{4}, BoundCnt: 5},
+			},
+			exp: sql.Histogram{
+				&Bucket{RowCnt: 5, DistinctCnt: 10, BoundVal: sql.Row{2}, BoundCnt: 5},
+				&Bucket{RowCnt: 10, DistinctCnt: 1, BoundVal: sql.Row{3}, BoundCnt: 10},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{4}, BoundCnt: 5},
+			},
+		},
+		{
+			inp: sql.Histogram{
+				&Bucket{RowCnt: 5, DistinctCnt: 10, BoundVal: sql.Row{2}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{2}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 5, BoundVal: sql.Row{4}, BoundCnt: 5},
+				&Bucket{RowCnt: 5, DistinctCnt: 1, BoundVal: sql.Row{4}, BoundCnt: 5},
+			},
+			exp: sql.Histogram{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{2}, BoundCnt: 10},
+				&Bucket{RowCnt: 10, DistinctCnt: 5, BoundVal: sql.Row{4}, BoundCnt: 10},
+			},
+		},
+	}
+	for i, tt := range tests {
+		t.Run(fmt.Sprintf("bin merge %d", i), func(t *testing.T) {
+			cmp, err := mergeOverlappingBuckets(tt.inp, []sql.Type{types.Int64, types.Int64})
+			require.NoError(t, err)
+			compareHist(t, tt.exp, cmp)
+		})
+	}
+}
+
+func TestEuclideanDistance(t *testing.T) {
+	tests := []struct {
+		x, y sql.Row
+		dist float64
+	}{
+		{
+			x:    sql.Row{0, 3},
+			y:    sql.Row{4, 0},
+			dist: 5,
+		},
+		{
+			x:    sql.Row{5, 0, 0},
+			y:    sql.Row{0, 12, 0},
+			dist: 13,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("%v x %v = %.2f", tt.x, tt.y, tt.dist), func(t *testing.T) {
+			cmp, err := euclideanDistance(tt.x, tt.y)
+			require.NoError(t, err)
+			require.Equal(t, tt.dist, cmp)
+		})
+	}
+}
+
+func TestBinAlignment(t *testing.T) {
+	tests := []struct {
+		left     []sql.HistogramBucket
+		right    []sql.HistogramBucket
+		expLeft  []sql.HistogramBucket
+		expRight []sql.HistogramBucket
+	}{
+		{
+			left: []sql.HistogramBucket{
```

Comment:
What determines how many buckets in the expected output? I was hoping the tests would make it clear but I'm not sure why, for instance, there's no 0 upper bound in the output of the second test.

Reply:
related to other comment, I don't know how to split (-infinity, 10) into (-infinity, 0) and (0,10). So I extend (-infinity, 0) to (-infinity, 10) by stealing (0,10) from (10,50).

Reply:
I'm not against tracking lower bound, i just didn't see that in other Dbs implementations so I was suspicious of whether it was practical. Pretty easy improvement later if so?

Reply:
I ended up having to implement lower bounds for the blog demo to work, it makes a big difference. On the Dolt side I'm going to leave the bound in-memory, storing it as a row with the rest of histograms was a bit awkward.